rust/storage/src/admission_controller.rs

```diff
@@ -0,0 +1,238 @@
+//! This is an admission controller suitable for deduping requests to object storage.
+//!
+//! It differs from the original admissioncontrolleds3 design in the following ways:
+//! - It does not directly store the future to be awaited in a map.  This pattern is broken because
+//!   it will call `poll` on the same future multiple times, which is not allowed by the contract
```

Comment:
This does not seem to be true. Shared futures are meant for this specific purpose and unless they are broken, the original impl should not be broken. I doubt that it is the case.

https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.shared
https://docs.rs/futures/latest/futures/future/struct.Shared.html